### PR TITLE
Method Giotto HMRF

### LINKED ISSUE #5: method_Giotto
**Update metadata:**
https://docs.google.com/spreadsheets/d/1QCeAF4yQG4bhZSGPQwwVBj_XF7ADY_2mK5xivAIfHsc/edit#gid=0

**Details of method (publication, GitHub):**
https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02286-2
https://github.com/drieslab/Giotto

### PR DESCRIPTION
**Pull request by shdam**
Giotto is a bit tricky to work with, but it is running now.
For inexplicable reasons, I could not get it to work with a provided PCA but only if running Giotto::runPCA.

Currently, it uses logcounts as the normalized data. Will there be more appropriate normalizations implemented or should I use the Giotto normalization instead?

Initially, I also implemented the non-spatial clustering methods in Giotto. But after the meeting last Wednesday, I decided to remove those.

Closes #5

### CODE REVIEW

**Comment by niklasmueboe** [comment]
What do you mean by more apropriate normalizations?
What would Giottos default workflow look like in terms of normalization?

File: method/Giotto/giotto_env.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (niklasmueboe) [inline method/Giotto/giotto_env.sh:1]: Change to `#!/usr/bin/env bash`

I think that's what we used everywhere else

File: method/Giotto/giotto_env.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+# Install the required R packages
+Rscript -e "remotes::install_version('colorRamp2', version = '0.1.0', repos = 'https://cran.r-project.org/')"
+Rscript -e "remotes::install_bitbucket(repo = 'qzhudfci/smfishhmrf-r', ref='2ab48253591b2dd3c545e117c4256f92ecb287ee')"
+# Install Giotto packages
+Rscript -e "remotes::install_github('drieslab/GiottoUtils', dependencies = FALSE, ref = '7c8f0010de6c916228834823455f48ed5b3fa706')"
+Rscript -e "remotes::install_github('drieslab/GiottoClass', dependencies = FALSE, ref = 'fca6eb3f5ee6e8e7e9cfe8a0bb82721107f4872d')"
+Rscript -e "remotes::install_github('drieslab/GiottoData', dependencies = FALSE, ref = '50606245a01f151c6c308f3282f7b3fd87c67027')"
+Rscript -e "remotes::install_github('drieslab/GiottoVisuals', dependencies = FALSE, ref = '8a68d8840ba4724b9a6cbc223dc7d6ef6f88f050')"
+Rscript -e "remotes::install_github('drieslab/Giotto', dependencies = FALSE, ref = 'fc7a6a51efc6853ff43f6028d1cce9a6070537e2')"
```
Comment (niklasmueboe) [inline method/Giotto/giotto_env.sh:17]: Is there a reason to use commit-hashes instead of releases?

File: method/Giotto/Giotto.r [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+HMRF_spatial_genes <- Giotto::doHMRF(
+    gobject = gobj,
+    spat_unit = "cell",
+    feat_type = "rna",
+    betas = c(0, 2, config$beta),
+    expression_values = "normalized",
+    spatial_genes = NULL, #my_spatial_genes,
+    dim_reduction_to_use = "pca",
+    dim_reduction_name = "PCA",
+    dimensions_to_use = 1:20,
```
Comment (niklasmueboe) [inline method/Giotto/Giotto.r:258]: If we cant use provided dimensionality reduction, maybe the number of dims should be a config parameter

**Review by niklasmueboe — CHANGES_REQUESTED** [review]

**Comment by shdam** [comment]
> What do you mean by more apropriate normalizations? What would Giottos default workflow look like in terms of normalization?

`gobj <- normalizeGiotto(gobj)`

> A. The standard method follows the standard protocol which can be adjusted using the provided parameters and follows the following order:
> 1. Data normalization for total library size and scaling by a custom scale-factor.
> 2. Log transformation of data.
> 3. Z-scoring of data by genes and/or cells.

It is commented out in line 221.

File: method/Giotto/giotto_env.sh (hunk 2 above)
Comment (shdam) [reply method/Giotto/giotto_env.sh:17]: Hmm, I actually didn't look too much into the difference, so mostly a force of habit. 
I did find this comment on a post-release version of Giotto:
> Update Matrix requirement to >= 1.6.2 (a re-install of irlba may resolve issues with Matrix incompatibility.)

So I will check if that can be fixed and irlba PCA can be used.

**Comment by niklasmueboe** [comment]
Hmm, I think as long as we are not providing PCA externally the data should be scaled before the PCA is run. But this might already be part of the PCA function. Can you check this?

**Comment by shdam** [comment]
> Hmm, I think as long as we are not providing PCA externally the data should be scaled before the PCA is run. But this might already be part of the PCA function. Can you check this?

runPCA takes in normalized data, but it is running with scale_unit = TRUE
> scale_unit | scale features before PCA (default = TRUE)

**Comment by niklasmueboe** [comment]
Than z-scoring is probably not needed, unless the gene expression info is used anywhere else in the workflow apart from PCA

**Comment by shdam** [comment]
It isn't. But is there a reason not to use their normalization method if we don't provide normalized counts other than log1p?

**Comment by niklasmueboe** [comment]
Well, the normalizations that we implement might be extended later. Also avoiding unnecessary computation if it is not used anyway

**Comment by shdam** [comment]
It seems hmrf might soon be updated:
[https://github.com/drieslab/Giotto/pull/877](https://github.com/drieslab/Giotto/pull/877)

**Review by niklasmueboe — APPROVED** [review]

**Comment by niklasmueboe** [comment]
I think there is a discussion to be had about using counts with feature selected data. Would we in that case limit feature selection also to methods only based on counts?

**Comment by shdam** [comment]
> I think there is a discussion to be had about using counts with feature selected data. Would we in that case limit feature selection also to methods only based on counts?

To add to the discussion, since Giotto has its own unique feature selection method, is it really Giotto's method if we feed it our own? - we could add a config of using either?

File: method/Giotto/Giotto.r [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+#my_spatial_genes <- km_spatialgenes[1:100]$feats
+HMRF_spatial_genes <- Giotto::doHMRF(
+    gobject = gobj,
+    spat_unit = "cell",
+    feat_type = "rna",
+    betas = c(0, 2, config$beta),
+    # expression_values = "normalized", # not used when dim_reduction_to_use is given
+    spatial_genes = my_spatial_genes,
+    dim_reduction_to_use = "pca",
+    dim_reduction_name = "PCA",
```
Comment (Jieran-S) [inline method/Giotto/Giotto.r:258]: Looking good! It makes sense to feed raw counts if they're using their own feature selection and transformation. Is `my_spatial_genes` defined here?

File: method/Giotto/giotto.yml
Comment (Jieran-S) [inline method/Giotto/giotto.yml:1]: Can we change all their name into `Giotto` for snakemake to properly process? (same as the folder name)

**Review by Jieran-S — CHANGES_REQUESTED** [review]

File: method/Giotto/Giotto.r (hunk 4 above)
Comment (shdam) [reply method/Giotto/Giotto.r]: Whoops, was meant to be `km_spatialgenes`

**Review by Jieran-S — APPROVED** [review]

**Pull request merged**